### PR TITLE
DOC: Clear up confusion between np.where(cond) and np.where(cond, x, y)

### PR DESCRIPTION
**Pull request by eric-wieser**
Eliminates all mentions of `np.where(cond)`, instead pointing the reader to np.nonzero.

Also:
* changes some example numbers to avoid collisions, making them easier to follow.
* adds some clearer examples for `where(cond, x, y)`

I think it's valuable not to muddle the documentation for these, as it leaves beginners confused about what `where` actually does.

### CODE REVIEW

File: numpy/add_newdocs.py [hunk 1]
```diff
@@ -1578,7 +1578,9 @@ def luf(lamdaexpr, *args, **kwargs):
 
     Return elements, either from `x` or `y`, depending on `condition`.
 
-    If only `condition` is given, return ``condition.nonzero()``.
+    The rest of this documentation covers only the case where all three
+    arguments are provided. When only `condition` is provided, this function
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: This was already true anyway - the `parameters` section did not reflect the `nonzero` case, so it was just odd that the return value did.

File: numpy/add_newdocs.py [hunk 2]
```diff
@@ -1591,56 +1593,53 @@ def luf(lamdaexpr, *args, **kwargs):
...
+            for c, xv, yv in zip(condition, x, y)
+        ]
 
     Examples
     --------
+    >>> a = np.arange(10)
+    >>> a
+    array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+    >>> np.where(a < 5, a, 10*a)
+    array([ 0,  1,  2,  3,  4, 50, 60, 70, 80, 90])
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: Seemed like this needed a simple 1d example

File: numpy/add_newdocs.py [hunk 3]
```diff
@@ -1591,56 +1593,53 @@ def luf(lamdaexpr, *args, **kwargs):
...
-    >>> np.where( x > 5 )
-    (array([2, 2, 2]), array([0, 1, 2]))
-    >>> x[np.where( x > 3.0 )]               # Note: result is 1D.
-    array([ 4.,  5.,  6.,  7.,  8.])
-    >>> np.where(x < 5, x, -1)               # Note: broadcasting.
-    array([[ 0.,  1.,  2.],
-           [ 3.,  4., -1.],
-           [-1., -1., -1.]])
-
-    Find the indices of elements of `x` that are in `goodvalues`.
```
Comment (eric-wieser) [inline numpy/add_newdocs.py]: This didn't really belong here - I put a similar remark in `isin` instead.

File: numpy/core/fromnumeric.py [hunk 4]
```diff
@@ -1615,16 +1615,16 @@ def nonzero(a):
 
     Examples
     --------
-    >>> x = np.array([[1,0,0], [0,2,0], [1,1,0]])
+    >>> x = np.array([[3, 0, 0], [0, 4, 0], [5, 6, 0]])
```
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: Changing these valeus makes them not be the same as the indices, which made things hard to follow

File: numpy/add_newdocs.py [hunk 5]
```diff
@@ -1591,56 +1593,53 @@ def luf(lamdaexpr, *args, **kwargs):
...
     nonzero, choose
 
     Notes
     -----
-    If `x` and `y` are given and input arrays are 1-D, `where` is
-    equivalent to::
+    If all the arrays are 1-D, `where` is equivalent to::
 
-        [xv if c else yv for (c,xv,yv) in zip(condition,x,y)]
+        [
```
Comment (mhvk) [inline numpy/add_newdocs.py]: Nitpicky, but I usually write this as
```
[(xv if c else yv)
 for c, xv, yv in zip(condition, x, y)]
```
I feel one looses the list-comprehension aspect by moving the `[` and `]` away from the code.

**Review by mhvk — APPROVED** [review]
Nice, much clearer. Had only a single, nitpicky comment.

File: numpy/add_newdocs.py (hunk 5 above)
Comment (eric-wieser) [reply numpy/add_newdocs.py]: I've never seen it written like that before. I'd probably drop the parens, but the wrapping seems unobjectionable to me.

**Comment by mattip** [comment]
Would be nice to add a similar remark to the beginning of `numpy.ma.where`

**Comment by eric-wieser** [comment]
Updated to be a little more obvious using `..note`.

Nits addressed too, I think

**Comment by eric-wieser** [comment]
I'd now contend that `where(c)` is actually harmful:

```python
>>> a = np.ma.array([1, 2, 3], mask=[True, False, False])
>>> np.nonzero(a)  #calls a.nonzero()
(array([1, 2], dtype=int64),)
>>> np.where(a)  # uh oh, calls `PyArray_NonZero`
(array([0, 1, 2], dtype=int64),)
```

File: numpy/ma/core.py [hunk 6]
```diff
@@ -7115,32 +7115,32 @@ def size(obj, axis=None):
...
+    Return a masked array with elements from `x` or `y`, depending on condition.
 
-    Returns a masked array, shaped like condition, where the elements
-    are from `x` when `condition` is True, and from `y` otherwise.
-    If neither `x` nor `y` are given, the function returns a tuple of
-    indices where `condition` is True (the result of
-    ``condition.nonzero()``).
+    .. note::
+        When only `condition` is provided, this function is identical to
+        `nonzero`. The rest of this documentation covers only the case where
```
Comment (eric-wieser) [inline numpy/ma/core.py]: `ma.where(c)` is safe, unlike `np.where(c)` which isn't

**Review by mhvk — APPROVED** [review]
Looks all OK. I'll merge.
